Add conversion of 3D label maps to labeled meshes with SurfaceNets.

jmargeta opened the pull request:
### Overview
This PR adds support for conversion of 3D label maps to labeled surfaces for colored visualization using vtkSurfaceNets3D algorithm from VTK. This is particularly useful for visualization of anatomical structures in medical images from their labels or the outputs of multi-label segmentation maps.

Resolves https://github.com/pyvista/pyvista/issues/5170

### Details
- added contour_labeled to ImageData -  contours can be extracted with `label_map.contour_labeled()`
- added example into the gallery with the frog tissue labels and a simple test for the number of extracted labels in the output mesh
![image](https://github.com/pyvista/pyvista/assets/1239359/4a8be7c7-7c20-4630-8690-cd826b8c7031)
- Note, vtk 9.3+ must be used which is currently in the pre-release stage on PyPI. Otherwise, the support with throw an exception. The example will, however, not be generated, and the test will be blank.


As mentioned in the issue, another appealing example could be made using the great [Neuroparc project](https://github.com/neurodata/neuroparc). Is it acceptable to make an example with an external dataset?
```python
import pooch
import pyvista as pv
file_name = pooch.retrieve(
    "https://github.com/neurodata/neuroparc/raw/master/atlases/label/Human/Schaefer1000_space-MNI152NLin6_res-1x1x1.nii.gz",
    known_hash="c3efe797aab3b3d9e705645bf29fac4e932c88dbe54ccaeb03982f11e66b3249",
)
label_map = pv.read(file_name)
label_map.contour_labeled(smoothing=True).plot(cmap="glasbey_bw")
```
![image](https://github.com/pyvista/pyvista/assets/1239359/00ce2f52-8297-498e-9a30-83a8add16ed2)

Does the name `contour_labeled` fit well with PyVista, or would you suggest a better one?

Also, I added types to the function, and had to silence these Mypy errors (as in https://github.com/pyvista/pyvista/pull/3837): 
```
pyvista/core/filters/image_data.py:853: error: Argument 1 to "set_default_active_scalars" has incompatible type "ImageDataFilters"; expected "DataSet"  [arg-type]
pyvista/core/filters/image_data.py:854: error: "ImageDataFilters" has no attribute "active_scalars_info"  [attr-defined]
pyvista/core/filters/image_data.py:858: error: "ImageDataFilters" has no attribute "get_array_association"  [attr-defined]
```

More info about the algorithm: Sarah F. Frisken, SurfaceNets for Multi-Label Segmentations with Preservation of Sharp Boundaries, Journal of Computer Graphics Techniques (JCGT), vol. 11, no. 1, 34-54, 2022. Available online http://jcgt.org/published/0011/01/03/

user27182 commented:
Awesome feature! Thanks for your contribution. This should be linked with #4802 in terms of PyVista VTK 9.3 support. `PackLabels` is used internally by `SurfaceNets`.

jmargeta commented:
Thanks, @user27182, for the pointer and for linking the two as it should be! Great to see the support for `PackLabels`.
Once your PR gets merged, I will be happy to adopt your vtk 9.3 detection logic.
Cannot wait to see the latest capabilities of vtk in PyVista.

MatthewFlamm commented:
vtk 9.3 is now in the CI, so it can be tested directly

MatthewFlamm commented on pyvista/core/filters/image_data.py:
```diff
@@ -772,3 +773,114 @@ def _flip_uniform(self, axis) -> 'pyvista.ImageData':
+        Examples
+        --------
+        See :ref:`contouring_example` for a full example using this filter.
+        """
```
Our standard is to have a space at the end.
```suggestion

        """
```

MatthewFlamm commented on pyvista/core/filters/image_data.py:
```diff
@@ -772,3 +773,114 @@ def _flip_uniform(self, axis) -> 'pyvista.ImageData':
+        if not hasattr(_vtk, 'vtkSurfaceNets3D'):  # pragma: no cover
+            from pyvista.core.errors import VTKVersionError
+
+            raise VTKVersionError('Surface nets 3D require VTK 9.3.20230807rc0 or newer.')
```
```suggestion
            raise VTKVersionError('Surface nets 3D require VTK 9.3.0 or newer.')
```

MatthewFlamm commented on examples/01-filter/contouring.py:
```diff
@@ -74,3 +74,15 @@
+# +++++++++++++++++++++++++
+#
+# Create labeled surfaces from 3D label maps (e.f. multi-label image segmentation)
+# using :func:`contour_labeled() <pyvista.PolyDataFilters.contour_labeled>`.
```
```suggestion
# using :func:`contour_labeled() <pyvista.ImageDataFilters.contour_labeled>`.
```

MatthewFlamm commented on pyvista/core/filters/image_data.py:
```diff
@@ -772,3 +773,114 @@ def _flip_uniform(self, axis) -> 'pyvista.ImageData':
+
+        This filter requires that the :class:`ImageData` has integer point
+        scalars, such as multi-label maps generated from image segmentation.
+
```
```suggestion
```

MatthewFlamm commented on pyvista/core/filters/image_data.py:
```diff
@@ -772,3 +773,114 @@ def _flip_uniform(self, axis) -> 'pyvista.ImageData':
+        if scalars is None:
+            set_default_active_scalars(self)  # type: ignore
+            field, scalars = self.active_scalars_info  # type: ignore
+            if field.value == 1:
```
Use the enums from 

https://github.com/pyvista/pyvista/blob/94c484c6a6a54b86704168a54666323234d7fdf5/pyvista/core/utilities/arrays.py#L15-L21

MatthewFlamm replied:
We should also use `field != FieldAssociation.POINT`, or similar.  It is probably certain that the only other possible value is cells here, but better to be safe and it is more understandable in the code.

jmargeta replied:
Thanks for highligting this enum for me, fully agree!

I copied the logic from gaussian_smooth and image_dilate_erode in the same file. Shall I update these two functions as well with the comparison against the enum for consistency (despite it being unrelated to this PR)?

jmargeta replied:
This is indeed ballooning quite a bit (9 test failures). I will not add it to this PR.

MatthewFlamm commented on pyvista/core/filters/image_data.py:
```diff
@@ -772,3 +773,114 @@ def _flip_uniform(self, axis) -> 'pyvista.ImageData':
+                raise ValueError('If `scalars` not given, active scalars must be point array.')
+        else:
+            field = self.get_array_association(scalars, preference='point')  # type: ignore
+            if field.value == 1:
```
same as above

jmargeta replied:
Cool, I also updated the error to give the detail about the given type (rather than saying "cell data" no matter what is provided):
`f'Can only process point data, given `scalars` are {field.name.lower()} data.'`

jmargeta commented:
@MatthewFlamm, thanks for the update, that is great to hear!

I added extra tests to improve the coverage and benefit from vtk 9.3 in the CI; and consider this wrapper complete functionally-wise.

Any feedback on the PR and what else I can do here to make the API better, please?
```python
def contour_labeled(
    self,
    n_labels: Optional[int] = None,
    smoothing: bool = False,
    smoothing_num_iterations: int = 50,
    smoothing_relaxation_factor: float = 0.5,
    smoothing_constraint_distance: float = 1,
    output_mesh_type: str = 'quads',
    output_style: str = 'default',
    scalars: Optional[str] = None,
    progress_bar: bool = False,
) -> 'pyvista.PolyData'
```

In particular, does the `contour_labeled` name align well with the spirit of PyVista? I chose it primarily for easier discoverability under the contour«TAB»-completion.

More explicit alternatives could be something like  `segmentaton`, `label_contours`, `label_map`, `surface_nets`, `label`, `label_surfaces`, `extract_label_map`, `segmentation_contours`... 

@tkoyama010, any chance for this to get into the November Release (#5138)?

jmargeta commented:
@MatthewFlamm, thanks a lot for the update on the CI and your review!
I believe to have addressed all your comments and consider the  wrapper complete functionally-wise.

Any feedback on what else I can do here to make the API better, please? In particular, does the `contour_labeled` name align well with the spirit of PyVista? I chose it primarily for easier discoverability under the contour«TAB»-completion.
```python
def contour_labeled(
    self,
    n_labels: Optional[int] = None,
    smoothing: bool = False,
    smoothing_num_iterations: int = 50,
    smoothing_relaxation_factor: float = 0.5,
    smoothing_constraint_distance: float = 1,
    output_mesh_type: str = 'quads',
    output_style: str = 'default',
    scalars: Optional[str] = None,
    progress_bar: bool = False,
) -> 'pyvista.PolyData'
```

More explicit alternatives could be something like  `segmentaton`, `label_contours`, `label_map`, `surface_nets`, `label`, `label_surfaces`, `extract_label_map`, `segmentation_contours`... 

@tkoyama010, any chance for this to get into the November Release (#5138)?

tkoyama010 commented:
@jmargeta I added this PR to the waiting list in the release. Let's add this to the next release!

jmargeta commented:
Awesome, thanks @tkoyama010 !

banesullivan commented on pyvista/core/filters/image_data.py:
```diff
@@ -772,3 +773,116 @@ def _flip_uniform(self, axis) -> 'pyvista.ImageData':
+        smoothing_constraint_distance : float, default: 1
+            Constraint distance of the smoothing.
+
+        output_mesh_type : str, default: 'quads'
+            Mesh type - triangles or quads.
```
I'm wondering if this should be an enum instead of a string

banesullivan replied:
Sounds good to me! Nice suggestion!

banesullivan commented on pyvista/core/filters/image_data.py:
```diff
@@ -772,3 +773,116 @@ def _flip_uniform(self, axis) -> 'pyvista.ImageData':
+        if not hasattr(_vtk, 'vtkSurfaceNets3D'):  # pragma: no cover
+            from pyvista.core.errors import VTKVersionError
+
+            raise VTKVersionError('Surface nets 3D require VTK 9.3.0 or newer.')
```
Can you add an `.. note::` in the docstring that this is only available for VTK 9.3+?

jmargeta replied:
Done.

banesullivan reviewed (commented):
Really awesome work! Happy to approve this, but I'm wondering if you can bump up the code coverage in the tests since some of the conditional options are not currently covered.

Otherwise small comments

jmargeta commented:
How about the name of the function itself? `labels.contour_labeled()` works well.
However, after using this internally for a while, it feels that `labels.surfaces_labeled()` or just `labels.surfaces()` would be slightly more ergonomic alternatives.

'contour' suggests computation of an iso-contour. Here, 'labels'  are a label map as an output from an image segmentation algorithm (3D image with discrete labels per voxel) we want to extract a surface "per label" rather than an iso-contour.
What do you think? Or is the original name fine?

banesullivan commented:
> How about the name of the function itself? labels.contour_labeled() works well.

+1 for `contour_labeled` as that is consistent with the few other contouring methods and makes it a bit more discovered with tab-complete. see also: `contour_banded()`

In my mind "contouring" is simply a geometry reducing operation: volume -> surface, surface -> line, line -> point. Since this is taking a volume (ImageData) and producing surfaces (PolyData), this naming feels appropriate to me.

Otherwise, since this is a direct wrapping of `vtkSurfaceNets3D` I don't think anyone could argue against `surface_nets()`

jmargeta commented:
@banesullivan, thank you very much for your thoughts. `contour_labeled` it is, and I am fairly happy with the PR now.
Please, do not hesitate to ping me if you see anything else to add/change.

banesullivan reviewed (approved):
LGTM! Thanks for your contributions, @jmargeta!

eigenvivek commented:
Thank you for this PR, it is very very helpful!

When I use `contour_labeled`, it prints some internal messages from VTK. Is there a way to suppress this by default? I don't want this showing up on a docs page I'm building. Also, messages like this don't show up when you use marching cubes, for example.

Thank you very much for your work, have been looking forward to the possibility of using SurfaceNets in PyVista for a few months now!

<img width="921" alt="Screen Shot 2023-11-25 at 11 33 28" src="https://github.com/pyvista/pyvista/assets/29757116/b38a6508-f38a-4552-9772-84adbbe50922">

banesullivan commented:
@eigenvivek, can you share a `pv.Report()`? I mainly want to know what your VTK version is

jmargeta commented:
Hi @eigenvivek,
Great to hear you are using this functionality!

You can try to change the vtk logging level before the contour_labeled execution. For example, this mutes the messages for me:
```python
import vtk
vtk.vtkLogger.SetStderrVerbosity(vtk.vtkLogger.ConvertToVerbosity(-1))
```
Disclaimer, I am not sure, this is the official way to do it :)

jmargeta commented:
All this is happening within the compiled part of VTK.
[  vtkLog(INFO, "Executing Surface Nets 3D");](https://github.com/Kitware/VTK/blob/2638973711fc9284ef71cadfb8468cb78ca427a6/Filters/Core/vtkSurfaceNets3D.cxx#L2322C1-L2322C1)

eigenvivek commented:
@jmargeta thank you for the suggestion, that worked perfectly :)

@banesullivan using `vtk==9.3.0`

<img width="482" alt="Screen Shot 2023-11-25 at 14 10 40" src="https://github.com/pyvista/pyvista/assets/29757116/b9431ea5-862d-47f8-8efb-52d5641d3233">

banesullivan commented:
Glad you got it fixed! FYI, this discussion also has some tips for suppressing these logs: https://github.com/pyvista/pyvista/discussions/3495

@jmargeta and @eigenvivek, if either of you think we should perform this suppression in this filter directly, don't hesitate to add that

jmargeta commented:
Hi @banesullivan, thanks for the pointer, `vtk.vtkLogger.SetStderrVerbosity(vtk.vtkLogger.VERBOSITY_OFF)` from the discussion you mentioned looks good. PR incoming.
